content/en/blog/_posts/2024-09-17-devsecops-in-action-k8s-secrets/index.md

```diff
@@ -0,0 +1,554 @@
+```go
+func httpHandle(w http.ResponseWriter, r *http.Request) {
+
+	if !changed {
+		if envVal, isExistEnv = os.LookupEnv(envVar); isExistEnv {
+			if filePath, isExistFile = fileExist(envVal); isExistFile {
+				body, err := os.ReadFile(filePath)
+				if err != nil {
+					log.Fatalf("unable to read file: %v", err)
+				}
+				os.Setenv(envVar, string(body))
+				envVal = os.Getenv(envVar)
+				changed = true
+			}
+		}
+	}
+
+	switch r.URL.Path {
+	case "/":
+		if isExistFile {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, envVal)
+		} else if isExistEnv {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, envVal)
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintln(w, "Secret Not Found")
+		}
+	case "/readiness":
+		if isExistEnv && isExistFile {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, "OK")
+		} else if isExistEnv && isExistFile == false && strings.HasPrefix(envVal, filePrefix) {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, "Environment variable was defined but file not found")
+		} else if isExistEnv {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, "Environment variable was defined")
+		} else {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintln(w, "Environment variable wasn't defined")
+		}
+	default:
+		w.WriteHeader(http.StatusNotFound)
+		fmt.Fprintln(w, "404 Not Found")
+	}
+}
+```
+In fact everything works and you can stop here. Unfortunately in the real world
+there are a lot of already working services and they must have implementation of
+additional environment variables. This isn't always possible and this approach
+can have specific limitations. In order to reduce the amount of limitations it's
+necessary to implement an approach based on support of both
+functions (variables) but without additional environment variables.
+This approach allows not to implement additional variable but to use the current
+variable for two cases:
+
+* legacy: using env variable for storing sensitive data
+* new: using env variable for storing path to file with sensitive data and
+override current value to value from the file
+
+To distinguish sensitive data values from path to file, by using prefixes,
+this will allow you to use the same variable and follow OWASP recommendations.
+
+First let's define variables and constants:
+
+[`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L12-L20)
+```go
+const (
+	envVar = "DEMO_SECRET__PASSWD"
+	filePrefix = "file://"
+)
+
+var (
+	envVal, filePath string
+	isExistFile, isExistEnv, changed bool
+)
+```
+
+Next, it's necessary to change the `fileExist` function. It's supposed to return
+a path to the file without a prefix if the file exists.
+
+[`cmd/main.go`](https://github.com/efrikin/devsecops-in-action-kubernetes-secrets/blob/8c8177b4955ce9cd9686c2746e5e2a810b9c2f62/cmd/main.go#L31-L43)
```

Comment:
We'll need to find a different home for this code; [for publication in this blog] it can't be your own repository.

Reply:
Fixed